tracker/peerstore/local.go

```diff
@@ -0,0 +1,274 @@
+	s.mu.RLock()
+	g, ok := s.peerGroups[h]
+	s.mu.RUnlock()
+	if !ok {
+		return nil, nil
+	}
+
+	g.mu.RLock()
+	defer g.mu.RUnlock()
+
+	if len(g.peerList) < n {
+		n = len(g.peerList)
+	}
+	if n <= 0 {
+		return nil, nil
+	}
+
+	result := make([]*core.PeerInfo, 0, n)
+
+	// Select n random indexes.
+	indexes := rand.Perm(len(g.peerList))
+	indexes = indexes[:n]
+
+	for _, i := range indexes {
+		// Note, we elect to return slightly expired entries rather than iterate
+		// until we find n valid entries.
+		e := g.peerList[i]
+		result = append(result, core.NewPeerInfo(e.id, e.ip, e.port, false /* origin */, e.complete))
+	}
+	return result, nil
+}
+
+// UpdatePeer implements Store.
+func (s *LocalStore) UpdatePeer(h core.InfoHash, p *core.PeerInfo) error {
+	g := s.getOrInitLockedPeerGroup(h)
+	defer g.mu.Unlock()
+
+	e, ok := g.peerMap[p.PeerID]
+	if !ok {
+		e = &peerEntry{}
+		g.peerList = append(g.peerList, e)
+		g.peerMap[p.PeerID] = e
+	}
+	e.id = p.PeerID
+	e.ip = p.IP
+	e.port = p.Port
+	e.complete = p.Complete
+	e.expiresAt = s.clk.Now().Add(s.config.TTL)
+
+	// Allows cleanupExpiredPeerGroups to quickly determine when the last
+	// peerEntry expires.
+	g.lastExpiresAt = e.expiresAt
+
+	return nil
+}
+
+func (s *LocalStore) getOrInitLockedPeerGroup(h core.InfoHash) *peerGroup {
+	// We must take care to handle a race condition against
+	// cleanupExpiredPeerGroups. Consider two goroutines, A and B, where A
+	// executes getOrInitLockedPeerGroup and B executes
+	// cleanupExpiredPeerGroups:
+	//
+	// A: locks s.mu, reads g from s.peerGroups, unlocks s.mu
+	// B: locks s.mu, locks g.mu, deletes g from s.peerGroups, unlocks g.mu
+	// A: locks g.mu
+	//
+	// At this point, A is holding onto a peerGroup reference which has been
+	// deleted from the peerGroups map, and thus has no choice but to attempt to
+	// reload a new peerGroup. Since the cleanup interval is quite large, it is
+	// *extremely* unlikely this for-loop will execute more than twice.
+	for {
+		s.mu.Lock()
+		g, ok := s.peerGroups[h]
+		if !ok {
+			g = &peerGroup{
+				peerMap:       make(map[core.PeerID]*peerEntry),
+				lastExpiresAt: s.clk.Now().Add(s.config.TTL),
+			}
+			s.peerGroups[h] = g
+		}
+		s.mu.Unlock()
+
+		g.mu.Lock()
+		if g.deleted {
+			g.mu.Unlock()
+			continue
+		}
+		return g
+	}
+}
+
+func (s *LocalStore) cleanupTask() {
+	for {
+		select {
+		case <-s.cleanupExpiredPeerEntriesTicker.C:
+			s.cleanupExpiredPeerEntries()
+		case <-s.cleanupExpiredPeerGroupsTicker.C:
+			s.cleanupExpiredPeerGroups()
+		case <-s.stop:
+			return
+		}
+	}
+}
+
+func (s *LocalStore) cleanupExpiredPeerEntries() {
+	s.mu.RLock()
+	hashes := make([]core.InfoHash, 0, len(s.peerGroups))
+	for h := range s.peerGroups {
+		hashes = append(hashes, h)
+	}
+	s.mu.RUnlock()
+
+	for _, h := range hashes {
+		s.mu.RLock()
+		g, ok := s.peerGroups[h]
+		s.mu.RUnlock()
+		if !ok {
+			continue
+		}
+
+		var expired []int
+		g.mu.RLock()
```

Comment:
You could actually consider just not locking here at all. If any entries are stale, they'll be checked again (with the lock) before they're deleted.

The benefit to this is you reduce the number of lock operations. This is nice because a lock operation always requires a memory access (it can't use the processor cache). While we don't really care about the extra cycles this introduces to the GC thread, if there are threads running on other processors (e.g. readers) that are also performing lock operations, there will be contention for the memory bus which may slow down the readers.

Totally optional

Reply:
The UpdatePeers function is appending to the slice, therefore it's not safe to access without a lock.